### PR TITLE
GenUI-ADO-2627: Data access request updates for Challenge Participants

### PR DESCRIPTION
**Pull request by grayson073**
## ADO Ticket:
[FEATURE 2627](https://dev.azure.com/GenUIADO/GenUI%20Sage%20Bionetworks%20-%20Challenge%20Portal/_workitems/edit/2627)

## Description:
Updates the `AccessRequirementsList` and related code to accept a `teamId` which in turn, gets used to fetch team requirements.

## Testing:
- Spin up component in Storybook and confirm that regular `AccessRequirementList` stories behave as accepted (not signed into a Synapse account, etc.) then;
- In `AccessRequirementList.tsx`:
  - Set prop `teamId` to default value of `'3445884'`
  - Set prop `requestObjectName` to default value of `'Awesome challenge'`
- Spin up component in Storybook and confirm screenshots below for not accepted yet, accepted terms, etc. 

## Screenshots:
Figma design:
<br>
Screenshot 2023-05-12 at 11 24 27 AM

<br><br><br>

Prior to accepting terms of the "Awesome Challenge" that are fetched from [team access requirements endpoint](https://rest-docs.synapse.org/rest/GET/team/id/accessRequirement.html):
<br>
Screenshot 2023-05-12 at 10 14 13 AM
<br><br>

After acceptance of terms with "View Terms" visible:
<br>
Screenshot 2023-05-12 at 10 34 21 AM
<br><br>

After acceptance of terms with "Hide Terms" visible:
Screenshot 2023-05-12 at 10 34 26 AM

### CODE REVIEW

**Comment by grayson073** [comment]
@nickgros I'm happy to add tests but given the switch to `vitest` (and possibly some other major switches happening right now?) I wasn't sure if they should be added/updated now. Would it be better to make a separate ticket for tracking that we can circle back to once the switch is complete?

File: packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx [hunk 1]
```diff
@@ -42,10 +43,12 @@ import { useSynapseContext } from '../../utils/SynapseContext'
 
 export type AccessRequirementListProps = {
   entityId: string // will show this entity info
+  teamId?: string // will show this team info
   accessRequirementFromProps?: Array<AccessRequirement>
   onHide: () => void
   renderAsModal?: boolean
   numberOfFilesAffected?: number // if provided, will show this instead of the entity information
+  requestObjectName?: string // if provided, will show this instead of the entity information or numberOfFilesAffected
```
Comment (grayson073) [inline packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:51]: image

File: packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx [hunk 2]
```diff
@@ -224,19 +258,10 @@ export default function AccessRequirementList(
 
       <DialogContent>
         <DialogSubsectionHeader sx={{ mt: 0 }}>
-          Access For:
+          What is this request for?
```
Comment (grayson073) [inline packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:261]: image

**Comment by nickgros** [comment]
@grayson073 I think it would be fine to add tests now. We haven't started the work to move synapse-react-client to vitest yet, and its API is almost identical to Jest's API, so the test logic shouldn't need to change

File: packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx [hunk 3]
```diff
@@ -141,12 +146,19 @@ export default function AccessRequirementList(
...
+          enabled: Boolean(!accessRequirementFromProps && teamId),
+        })
+      : useGetAccessRequirementsForEntity(entityId, {
+          enabled: Boolean(!accessRequirementFromProps && !teamId && entityId),
+        })
+
+    return fetched || {}
+  }
+
+  const { data: fetchedRequirements } = determineFetchedARs()
```
Comment (nickgros) [inline packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:161]: This is breaking the rules of hooks (indicated in the lint job failure).

Maybe you could do something like this?

```suggestion
  const { data: fetchedRequirementsForTeam } = useGetAccessRequirementsForTeam(
    entityId,
    {
      enabled: Boolean(!accessRequirementFromProps && teamId),
    },
  )
  const { data: fetchedRequirementsForEntity } = useGetAccessRequirementsForEntity(
    entityId,
    {
      enabled: Boolean(!accessRequirementFromProps && entityId && !teamId),
    },
  )

  const fetchedRequirements = teamId ? fetchedRequirementsForTeam : fetchedRequirementsForEntity
```

Comment (grayson073) [reply packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:161]: Good catch, thank you. I had similar to suggested originally but shortened things, only to realize my linter isn't catching correctly (Intellij IDEA). I forced a bad use of `useMemo` with a hook inside as well to test but there seems to be some inconsistencies in warnings.

Do you happen to be pointing your IDE to a specific eslint config within `synapse-react-client` or letting it automatically scan the project?

Comment (grayson073) [reply packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx]: Sorry I meant within `synapse-web-monorepo`

File: packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx [hunk 4]
```diff
@@ -141,13 +146,21 @@ export default function AccessRequirementList(
 
   const entityInformation = useGetInfoFromIds<EntityHeader>(entityHeaderProps)
 
-  const { data: fetchedRequirements } = useGetAccessRequirementsForEntity(
+  const { data: fetchedRequirementsForTeam } = useGetAccessRequirementsForTeam(
     entityId,
```
Comment (grayson073) [inline packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:150]:
```suggestion
    teamId,
```

File: packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx [hunk 5]
```diff
@@ -141,13 +146,21 @@ export default function AccessRequirementList(
 
   const entityInformation = useGetInfoFromIds<EntityHeader>(entityHeaderProps)
 
-  const { data: fetchedRequirements } = useGetAccessRequirementsForEntity(
+  const { data: fetchedRequirementsForTeam } = useGetAccessRequirementsForTeam(
+    teamId,
```
Comment (grayson073) [inline packages/synapse-react-client/src/lib/containers/AccessRequirementList/AccessRequirementList.tsx:150]:
```suggestion
    teamId!,
```

**Review by nickgros — COMMENTED** [review]
LGTM!

**Review by nickgros — APPROVED** [review]

**Pull request merged**